crypto/src/key/secp256k1/extended_keys.rs

```diff
@@ -0,0 +1,256 @@
+use crate::key::hdkd::chain_code::CHAINCODE_LENGTH;
+use crate::key::hdkd::{
+    chain_code::ChainCode,
+    child_number::ChildNumber,
+    derivable::{Derivable, DerivationError},
+};
+use crate::key::secp256k1::{Secp256k1PrivateKey, Secp256k1PublicKey};
+use crate::random::{CryptoRng, Rng};
+use hmac::{Hmac, Mac};
+use secp256k1;
+use secp256k1::SECP256K1;
+use serialization::{Decode, Encode};
+use sha2::Sha512;
+use zeroize::Zeroize;
+
+// Create alias for HMAC-SHA512
+type HmacSha512 = Hmac<Sha512>;
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct Secp256k1ExtendedPrivateKey {
+    /// Chain code
+    pub chain_code: ChainCode,
+    /// Private key
+    pub private_key: Secp256k1PrivateKey,
+}
+
+impl Secp256k1ExtendedPrivateKey {
+    pub fn new_master(seed: &[u8]) -> Result<Secp256k1ExtendedPrivateKey, DerivationError> {
+        let mut mac =
+            HmacSha512::new_from_slice(b"Bitcoin seed").expect("HMAC can take key of any size");
+        mac.update(seed);
+        let mut result = mac.finalize().into_bytes();
+
+        let private_key = secp256k1::SecretKey::from_slice(&result[..32])
+            .map_err(|_| DerivationError::KeyDerivationError)?
+            .into();
+
+        let chain_code: [u8; CHAINCODE_LENGTH] =
+            result[32..].try_into().expect("Chaincode size is 32 bytes");
+        let chain_code = ChainCode::from(chain_code);
+
+        result.zeroize();
+
+        Ok(Secp256k1ExtendedPrivateKey {
+            private_key,
+            chain_code,
+        })
+    }
+
+    pub fn new<R: Rng + CryptoRng>(
+        rng: &mut R,
+    ) -> (Secp256k1ExtendedPrivateKey, Secp256k1ExtendedPublicKey) {
+        // Create a new chain code
+        let mut chain_code = [0u8; 32];
+        rng.fill_bytes(&mut chain_code);
+        let chain_code = chain_code.into();
+        let private_key = secp256k1::SecretKey::new(rng).into();
+        // Generate a new private key
+        let ext_priv = Secp256k1ExtendedPrivateKey {
+            private_key,
+            chain_code,
+        };
+        // Generate the public key
+        let ext_pub = Secp256k1ExtendedPublicKey::from_private_key(&ext_priv);
+        // Return the pair
+        (ext_priv, ext_pub)
+    }
+
+    pub fn as_private_key(&self) -> &Secp256k1PrivateKey {
+        &self.private_key
+    }
+}
+
+impl Derivable for Secp256k1ExtendedPrivateKey {
+    fn derive_child(self, num: ChildNumber) -> Result<Self, DerivationError> {
+        let mut mac = HmacSha512::new_from_slice(&self.chain_code.into_array())
+            .expect("HMAC can take key of any size");
+
+        let secp_key = self.private_key.data;
+
+        // We only support hard derivations
+        if num.is_hardened() {
+            mac.update(&[0u8]);
+            mac.update(&secp_key[..]);
+        } else {
+            return Err(DerivationError::UnsupportedDerivationType);
+        }
+
+        // Add the child number
+        mac.update(&num.to_encoded_be_bytes());
```

Comment:
Please move all these stateful mac updates into a separate function. The rule of thumb is: Whenever we play with `mut`, we try to isolate it. It's not always possible, but we do it whenever we can, and it seems to be possible in your case.

So, do the check if `num` is hardened (which I'm not convinced is necessary, since like I said, it's not supported anyway at the `ChildNumber` level). I don't know why we're checking then appending `0u8` followed by `secp_key`. Does being hardened have any bearing on that? If yes, it would be nice to move any specifical part of "being hardened" into the `ChildNumber` struct, and just retrieve it with a method. I think we can at least take the `secp_key` appending out as it's neutral and doesn't matter.

Then what do we have left? We create a function that takes a ref to the private key data + a ref to num, and returns the result array. That would be beautifully more readable and separately testable.

Reply:
I moved the some of the functions to the `to_key_and_chain_code` function. I tried moving out the mut parts to a separate function but it makes the code harder to follow because it breaks in half the BIP32 algorithm.

I simplified the ChildNumber API so we don't have to deal with non-hardened paths. The `0u8` is needed for the private key padding and it is unrelated to the child number.